lib/spack/spack/test/cmd/uninstall.py

```diff
@@ -166,3 +169,192 @@ def _warn(*args, **kwargs):
+            # --remove was not specified
+            assert set(root.name for (root, _) in e1.concretized_specs()) == set(
+                ["dt-diamond-left", "dt-diamond-bottom"]
+            )
+
+            for _, concretized_spec in e1.concretized_specs():
+                assert not concretized_spec.package.installed
+
+        # Everything in e2 depended on dt-diamond-bottom, so should also
+        # have been uninstalled. The roots should be unchanged though.
+        e2 = spack.environment.read("e2")
+        with e2:
+            assert set(root.name for (root, _) in e2.concretized_specs()) == set(
+                ["dt-diamond-right", "dt-diamond-bottom"]
+            )
+            for _, concretized_spec in e2.concretized_specs():
+                assert not concretized_spec.package.installed
+
+    def test_uninstall_remove_dependency_shared_between_envs(self, environment_setup):
+        """If you "spack uninstall --dependents --remove dt-diamond-bottom" from
+        e1, then all packages are removed from e1 (it is now empty);
+        dt-diamond-left is also uninstalled (since only e1 needs it) but
+        dt-diamond-bottom is not uninstalled (since e2 needs it).
+        """
+        e1 = spack.environment.read("e1")
+        with e1:
+            dtdiamondleft = next(
+                concrete
+                for (_, concrete) in e1.concretized_specs()
+                if concrete.name == "dt-diamond-left"
+            )
+            output = uninstall("-y", "--dependents", "--remove", "dt-diamond-bottom")
+            assert "The following specs will be removed but not uninstalled" in output
+            assert not list(e1.roots())
+            assert not dtdiamondleft.package.installed
+
+        # Since -f was not specified, all specs in e2 should still be installed
+        # (and e2 should be unchanged)
+        e2 = spack.environment.read("e2")
+        with e2:
+            assert set(root.name for (root, _) in e2.concretized_specs()) == set(
+                ["dt-diamond-right", "dt-diamond-bottom"]
+            )
+            for _, concretized_spec in e2.concretized_specs():
+                assert concretized_spec.package.installed
+
+    def test_uninstall_dependency_shared_between_envs_fail(self, environment_setup):
+        """If you "spack uninstall --dependents dt-diamond-bottom" from
+        e1 (without --remove or -f), then this should fail (this is needed by
+        e2).
+        """
+        e1 = spack.environment.read("e1")
+        with e1:
+            # FAILURE: this attempts to emulate tests in cmd/install.py but
+            # according to logic in main.py, none of the output printed by
+            # a command will be included in the exception
+            with pytest.raises(
+                spack.error.SpackError,
+                match=r".*Will not uninstall.*There are still dependents.*use `spack env remove`",
+            ):
```

Comment:
@becker33 this is my current attempt at following https://github.com/spack/spack/pull/33711#discussion_r1015941410 , and something I'm doing is incorrect and I also am having trouble following the logic to see how this output would make it into the exception: I assumed it wouldn't based on reading `main.py:SpackCommand.__call__`, but as noted in this comment, there are tests which seem to be able to examine output.

I'll note one difference between this and `test/cmd/install.py:test_install_invalid_spec` is that the error is actually raised as an exception (via parse.py) which triggers:

```
            except BaseException as e:
                tty.debug(e)
                self.error = e
                if fail_on_error:
                    self._log_command_output(out)
                    raise
```

So that's my theory of why tests which  check for failures in commands typically run with `fail_on_error=False` (I originally copied this strategy from looking elsewhere, which is admittedly naive but on further inspection it would seem necessary pending a refactor of `SpackCommand`).